### DISCUSSION TITLE
Global variables in Rules and Berry Scripting

### DISCUSSION THREAD

**Discussion by indigmas** [arendst/Tasmota#20961] [Q&A] [answered]
Hello Community, is there any way to use global variables not only in tasmota scripting language but also in rules and in berrry scripting?

**Comment by barbudor** [comment, marked as answer]
Do you mean global variables in each of rules/scripting/berry independently or common to all of them ?

In rules, tasmota Var and Mem are 'global' (considering there is no 'loczl' ones) 
In Scripting, the '>D' section if I'm correct defines variables global to the script
In Berry, anything not in a class/object instance is global


1st, rules and scripting are exclusive features, you can't have global variables shared between them

Berry can access to Tasmota Var and Mem through commands and parsing the result so it can be used to échange between rules and Berry. Although I do 't see the point in splitting something between rules and Berry as you can do the same, better and more in Berry onky

AFAIK there is no way to échange simply between Scripting and Berry. Not impossible but may not be straight forward

**Reply by sfromis** [reply]
While there is no Berry/scripter integration, you can do things to have Berry code get/set a scripter variable etc using code like this:
`tasmota.cmd('script >=>br berryvar=%scriptvar%')`
`tasmota.cmd('script >scriptvar=345')`
`berryvar=tasmota.cmd('script ?scriptvar')['script']['scriptvar']`

This is using the command `script >` to execute a bit of scripter code, which can issue the `br` command with a bit of Berry code to assign a value to a Berry global variable. If could also do other stuff you could do in the Berry console, like call some function.

You should be able to use a similar approach to have rules get/set scripter variables, by using the `script >` command along with rules variables.

To be clear, what is "global" variables in Berry is about being in the global context, instead of being part of some instance, a very different thing than the scripter concept of global variables being shared across devices.

**Reply by barbudor** [reply]
Var/Mem/Rules are exclusive to Scripting-Language so they can't co-exist

**Reply by sfromis** [reply]
Yeah, forgot about rules/scripter being mutually exclusive.

If someone has the urge, it should be feasible to mimic the scripter feature of cross-device global shared variables using Berry, as the mechanism is based on UDP.

**Comment by indigmas** [comment]
ok, looks like i have to learn berry
